File: stations/hook-observatory/handlers/external.py
"""
External handler wrappers — subprocess calls to scripts that live outside hook-observatory.

These scripts are part of other systems (memvault, session-redactor, playwright).
They're wrapped here to maintain the unified handler interface while keeping
the source of truth in their respective repos.
"""

from __future__ import annotations

import os

from .base import HOME, HookResult, call_external_script

MEMVAULT_SCRIPTS = os.path.join(HOME, "workshop", "mcp", "memvault", "scripts")


def recall(event_type: str, tool_name: str, tool_input: dict, raw_input: str) -> HookResult:
    """UserPromptSubmit: memvault recall — returns markdown context."""
    return call_external_script(
        os.path.join(MEMVAULT_SCRIPTS, "recall_v2.py"),
        input_data=raw_input,
        timeout=15,
    )


def extract(event_type: str, tool_name: str, tool_input: dict, raw_input: str) -> HookResult:
    """SessionEnd: async memory extraction pipeline."""
    return call_external_script(
        os.path.join(MEMVAULT_SCRIPTS, "extract_async.py"),
        input_data=raw_input,
        timeout=5,
    )


def skill_tracker(event_type: str, tool_name: str, tool_input: dict, raw_input: str) -> HookResult:
    """PostToolUse/Skill: track skill usage to Core API."""
    return call_external_script(
        os.path.join(MEMVAULT_SCRIPTS, "skill_tracker_v2.py"),
        input_data=raw_input,
        timeout=5,
    )


def progressive_extract(
    event_type: str, tool_name: str, tool_input: dict, raw_input: str
) -> HookResult:
    """PreCompact: async progressive memory extraction — fire-and-forget."""
    import tempfile

    from .base import HOME, run_background

    python = os.path.join(HOME, ".local", "bin", "python3")

    # Write input to temp file for background process
    try:
        with tempfile.NamedTemporaryFile(
            mode="w", prefix="memvault-prog-", suffix=".json", dir="/tmp", delete=False
        ) as tmpf:
            tmpf.write(raw_input)
            tmpfile_path = tmpf.name
    except Exception:
        return HookResult()

    script = os.path.join(MEMVAULT_SCRIPTS, "extract_progressive.py")
    if not os.path.isfile(script):
        return HookResult()

    # Background: cat input | python extract_progressive.py
    try:
        run_background(
            f"cat {tmpfile_path} | {python} {script}; rm -f {tmpfile_path}",
        )
    except Exception:
        # Cleanup temp file if background dispatch fails
        try:
            os.unlink(tmpfile_path)
        except OSError:
            pass
    return HookResult()


def sync_login(event_type: str, tool_name: str, tool_input: dict, raw_input: str) -> HookResult:
    """SessionStart: sync Playwright profile from master."""
    return call_external_script(
        os.path.join(HOME, ".playwright-profiles", "sync-login.sh") + " --hook",
        input_data=raw_input,
        timeout=15,
    )

File: stations/hook-dispatcher/internal/handlers/external.go
package handlers

// external.go — Go port of handlers/external.py
//
// Wraps four sub-handlers that call external scripts:
//   - recall        → UserPromptSubmit  (memvault recall)
//   - skill_tracker → PostToolUse/Skill (skill usage tracker)
//   - progressive_extract → PreCompact  (fire-and-forget extraction)
//   - sync_login    → SessionStart      (Playwright profile sync)
//
// All are deferrable. Any error → fail-open (core.Allow()).

import (
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/joneshong/hook-dispatcher/internal/core"
)

func init() {
	core.Register("UserPromptSubmit", core.Entry{
		Matcher:    "",
		Handler:    externalRecall,
		Critical:   false,
		ModuleName: "external",
	})
	core.Register("PostToolUse", core.Entry{
		Matcher:    "Skill",
		Handler:    externalSkillTracker,
		Critical:   false,
		ModuleName: "external",
	})
	core.Register("PreCompact", core.Entry{
		Matcher:    "",
		Handler:    externalProgressiveExtract,
		Critical:   false,
		ModuleName: "external",
	})
	core.Register("SessionStart", core.Entry{
		Matcher:    "",
		Handler:    externalSyncLogin,
		Critical:   false,
		ModuleName: "external",
	})
}

func memvaultScripts() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, "workshop", "mcp", "memvault", "scripts")
}

// externalRecall runs the memvault recall.py script and returns its output
// as a UserPromptSubmit text injection.
func externalRecall(_, _ string, _ map[string]any, rawInput string) core.HookResult {
	script := filepath.Join(memvaultScripts(), "recall.py")
	python := pythonBin()
	r := core.RunCmd([]string{python, script}, rawInput, 15*time.Second, "")
	if r == nil || r.ExitCode != 0 {
		return core.Allow()
	}
	out := r.Stdout
	if out == "" {
		return core.Allow()
	}
	return core.TextResult(out)
}

// externalSkillTracker runs skill_tracker.py fire-and-forget style.
func externalSkillTracker(_, _ string, _ map[string]any, rawInput string) core.HookResult {
	script := filepath.Join(memvaultScripts(), "skill_tracker.py")
	python := pythonBin()
	if _, err := os.Stat(script); err != nil {
		return core.Allow()
	}
	_ = core.RunBackground([]string{
		"sh", "-c",
		fmt.Sprintf("echo %s | %s %s", shellQuote(rawInput), python, shellQuote(script)),
	}, "")
	return core.Allow()
}

// externalProgressiveExtract writes rawInput to a temp file and runs
// extract_progressive.py in the background (fire-and-forget, PreCompact).
func externalProgressiveExtract(_, _ string, _ map[string]any, rawInput string) core.HookResult {
	script := filepath.Join(memvaultScripts(), "extract_progressive.py")
	if _, err := os.Stat(script); err != nil {
		return core.Allow()
	}

	// Write rawInput to temp file — background process reads it.
	tmpf, err := os.CreateTemp("/tmp", "memvault-prog-*.json")
	if err != nil {
		return core.Allow()
	}
	tmpPath := tmpf.Name()
	if _, err := tmpf.WriteString(rawInput); err != nil {
		_ = tmpf.Close()
		_ = os.Remove(tmpPath)
		return core.Allow()
	}
	_ = tmpf.Close()

	python := pythonBin()
	_ = core.RunBackground([]string{
		"sh", "-c",
		fmt.Sprintf("cat %s | %s %s; rm -f %s",
			shellQuote(tmpPath), python, shellQuote(script), shellQuote(tmpPath)),
	}, "")

	return core.Allow()
}

// externalSyncLogin runs ~/.playwright-profiles/sync-login.sh --hook.
func externalSyncLogin(_, _ string, _ map[string]any, rawInput string) core.HookResult {
	home, _ := os.UserHomeDir()
	script := filepath.Join(home, ".playwright-profiles", "sync-login.sh")
	if _, err := os.Stat(script); err != nil {
		return core.Allow()
	}
	r := core.RunCmd([]string{script, "--hook"}, rawInput, 15*time.Second, "")
	if r == nil || r.ExitCode != 0 {
		return core.Allow()
	}
	return core.Allow()
}

// --- helpers -----------------------------------------------------------------

// pythonBin returns the canonical python3 path for this workshop.
func pythonBin() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".local", "bin", "python3")
}

// shellQuote wraps s in single quotes, escaping any embedded single quotes.
func shellQuote(s string) string {
	result := "'"
	for _, ch := range s {
		if ch == '\'' {
			result += "'\\''"
		} else {
			result += string(ch)
		}
	}
	result += "'"
	return result
}
